src/Microsoft.ML.LightGbm/LightGbmTrainerBase.cs

```diff
@@ -371,6 +373,9 @@ private protected override TModel TrainModelCore(TrainContext context)
             return CreatePredictor();
         }
 
+        private protected virtual void InitializeBeforeTraining()
+        { return; } // MYTODO: Is there a better way to avoid having to do this? An abstract method?
```

Comment:
So, originally the idea that @yaeldekel suggested to me offline was to create an abstract method `InitializeBeforeTraining`, and add the implementation in the `LightGbmMulticlassTrainer` child class to always initialize `_tlcNumClass` and `_numClass` when fitting.

Problem with that is that then I would need to provide an implementation also in the other children of `LightGbmTrainerBase`, but as far as issue #3878 concerns, there isn't any problem in those classes. So instead of making an abstract method, I made this virtual empty method that is overridden in the multiclass LightGBM trainer. Still, I am unsure if this is the best way to do this, as it seems ugly to me... but I wouldn't know better.

Reply:
A virtual method is good. You don't need to have `return;` in there though, you can just have empty braces: `{ }`.

---
In reply to: [362128656](https://github.com/dotnet/machinelearning/pull/4608#discussion_r362128656) [](ancestors = 362128656)